Explain scheduler fine-tuning better (#18356)
mistercrunch/Airflow: docs/apache-airflow/concepts/scheduler.rst

  Microsoft SQLServer has not been tested with HA.

Fine-tuning your Scheduler
--------------------------

When you deploy Airflow in production you often would like to optimize its performance and

> fine-tune Scheduler behaviour. Firs of all you need to remember that Scheduler performs two

Reviewer: ```suggestion
fine-tune Scheduler behaviour. First of all you need to remember that Scheduler performs two
```

> operations:

Reviewer 2: ```suggestion
The Scheduler is responsible for two operations:
```

> * continuously parses DAG files and updates their starts in ``Serialized DAG`` form in the database
> * continuously finds and schedules for execution the next tasks to run and sends those tasks for
>   execution to the executor you have configured

Reviewer 2: Nothing wrong here, but would try to avoid internal details not necessary to the story

```suggestion
1. continuously parsing DAG files and synchronizing with the DAG in the database
2. continuously scheduling tasks for execution
```

> Those two tasks are executed in parallel by scheduler, they are fairly independent from each other and
> they are run using different processes. You can fine tune the behaviour of both components, however
> in order to fine-tune your scheduler, you need to included a number of factors:

Reviewer 2: ```suggestion
Those two tasks are executed in parallel by the scheduler and run independently of each other in different processes. In order to fine-tune your scheduler, you need to include a number of factors:
```

* The kind of deployment you have

>     * what kind of filesystem you have to share the DAGS

Reviewer: ```suggestion
    * what kind of filesystem you have to share the DAGs
```

    * how fast the filesystem is (in many cases of distributed cloud filesystem you can pay extra to get
      more throughput/faster filesystem
    * how much memory you have for your processing
    * how much CPU you have available
    * how much networking throughput you have available

* The logic and definition of your DAG structure:
    * how many DAG files you have
    * how many DAGs you have in your files
    * how large the DAG files are (remember scheduler needs to read and parse the file every n seconds)
    * how complex they are
    * whether parsing your DAGs involves heavy processing (Hint! It should not. See:doc:`/best-practices`)

* The scheduler configuration
   * How many schedulers you have
   * How many parsing processes you have in your scheduler
   * How much time scheduler waits between re-parsing of the same DAG (it happens continuously)
   * How many task instances scheduler processes in one loop

>    * How many new dag runs should be created/scheduled per loop

Reviewer: ```suggestion
   * How many new DAG runs should be created/scheduled per loop
```

   * Whether to execute "mini-scheduler" after completed task to speed up scheduling dependent tasks
   * How often the scheduler should perform cleanup and check for orphaned tasks/adopting them
   * Whether scheduler uses row-level locking

Airflow gives you a lot of "knobs" to turn to fine tune the performance but it's a separate task,
depending on your particular deployment, your DAG structure, hardware availability and expectations,
to decide which knobs to turn to get best effect for you. Part of the job when managing the
deployment is to decide what you are going to optimize for. Some users are ok with

> 30 seconds delays of new DAG parsing, at the expense of lower CPU usage, where some other users

Reviewer: ```suggestion
30 seconds delays of new DAG parsing, at the expense of lower CPU usage, whereas some other users
```

expect the DAGs to be parsed almost instantly when they appear in the DAGs folder at the
expense of higher CPU usage for example.

Airflow gives you the flexibility to decide, but you should find out what aspect of performance is
most important for you and decide which knobs you want to turn in which direction.

Generally for fine-tuning, your approach should be the same as for any performance improvement and
optimizations (we will not recommend any specific tools - just use the tools that you usually use
to observe and monitor your systems):

* decide which aspect of performance is most important for you (what you want to improve)
* observe your system to see where your bottlenecks are: CPU, memory, I/O are the usual limiting factors
* based on your expectations and observations - decide what is your next improvement and go back to

>   the observation of your performance, bottlenecks. Performance improvement is an iterative process

Reviewer: ```suggestion
  the observation of your performance, bottlenecks. Performance improvement is an iterative process.
```

The improvements that you can consider are:

* improve utilization of your resources. This is when you have a free capacity in your system that
  seems underutilized (again CPU, memory I/O, networking are the prime candidates) - you can take
  actions like increasing number of schedulers, parsing processes or decreasing intervals for more
  frequent actions might bring improvements in performance at the expense of higher utilization of those.

> * increase hardware capacity (for example if you see that CPU is limiting you or tha I/O you use for

Reviewer: ```suggestion
* increase hardware capacity (for example if you see that CPU is limiting you or that I/O you use for
```

  DAG filesystem is at its limits). Often the problem with scheduler performance is
  simply because your system is not "capable" enough and this might be the only way. For example if
  you see that you are using all CPU you have on machine, you might want to add another scheduler on
  a new machine - in most cases, when you add 2nd or 3rd scheduler, the capacity of scheduling grows
  linearly (unless the shared database or filesystem is a bottleneck).
* experiment with different values for the "scheduler tunables". Often you might get better effects by
  simply exchanging one performance aspect for another. For example if you want to decrease the

>   cpu usage, you might increase file processing interval (but the result will be that new DAGs will

Reviewer: ```suggestion
  CPU usage, you might increase file processing interval (but the result will be that new DAGs will
```

  appear with bigger delay). Usually performance tuning is the art of balancing different aspects.
* sometimes you change scheduler behaviour slightly (for example change parsing sort order)
  in order to get better fine-tuned results for your particular deployment.

In order to perform fine-tuning, it's good to understand how Scheduler works under-the-hood.
You can take a look at the ``Airflow Summit 2021``
`Deep Dive into the Airflow Scheduler talk <https://youtu.be/DYC4-xElccE>`_ to perform the fine-tuning.

Here are the most important tunables you can use to impact various performance aspects of the Scheduler:

.. _scheduler:ha:tunables:

> Scheduler Tunables

Reviewer 2: Would keep this aligned with Airflow terminology
```suggestion
Scheduler configuration options
```

Author: Good for me . The "tunables" word sounds nice but is a bit strange indeed

""""""""""""""""""

The following config settings can be used to control aspects of the Scheduler.
However you can also look at other scheduler configuration parameters available at

> :doc:`../configurations-ref` in ``[scheduler]`` section.

Reviewer 2: Would leave this out, or mention something like "non-performance related configuration options"

  usually as fast as it can be, especially if your machines use fast SSD disks for local storage. Those
  distribution mechanisms have other characteristics that might make them not the best choice for you,
  but if your problems with performance come from distributed filesystem performance, they might be the
  best approach to follow.
* Database connections and Database usage might become a problem as you want to increase performance and
  process more things in parallel. Airflow is known from being "database-connection hungry" - the more DAGs
  you have and the more you want to process in parallel, the more database connections will be opened.
  This is generally not a problem for MySQL as its model of handling connections is thread-based, but this
  might be a problem for Postgres, where connection handling is process-based. It is a general consensus
  that if you have even medium size Postgres-based Airflow installation, the best solution is to use
  `PGBouncer <https://www.pgbouncer.org/>`_ as a proxy to your database. The :doc:`helm-chart:index`
  supports PGBouncer out-of-the-box. For MsSQL we have not yet worked out the best practices as support

>   for MsSQL is still experimental.

Reviewer 3: is MsSQL experimental?
I thought that starting 2.2 we consider it to be stable and fully supported

Author: When we release 2.2 it will not be. But I plan to merge this into v2-1 branch and publish it for 2.1.4

* CPU usage is most important for FileProcessors - those are the processes that parse and execute

>   Python DAG files. Since Schedulers triggers such parsing continuously, when you have a lot complex DAGs,

Reviewer 3: What is a complex DAG? do we mean just high number of tasks?

Author: I think there are multiple things that might be seen as "complex" , For example "one that takes a lot of time to parse", generates "plenty of dags", "generates plenty of tasks", "generates plenty of tasks interconnection". I have no (and I think no-one else has) a ready to give recipe "this is too complex" or "this is simple enough". 

Each user can write their dags in their own ways. We can determine some bad practices (and we did already - with external DB usage for example). Python is so versatile that it's difficult to say "this is complex" this is "simple". However it's rather easy to say "this is less complex than it's been before" when you simplify it" and I think users should be gently guided that they should understand what they do when they write Python code and work out their own "ways" of developing DAGs which lead to less complexity.

But this is a good point to point people to some ways of at least assessing it - we have the right chapter in `best practices` about testing dag - for example "check tha your DAG imports <fast>" and try to speed it up, if you perceive this being too long (again we cannot tell what "fast" is  - for some users this might be ms, for some several seconds is still ok. But every one of those users can be told `if you want to improve performance - made DAG loading "faster", and here is the tool to test it`. This is precisely what I am going to do.

I will describe the above. I think users should be made aware that there is no ready-to-use-recipe for simple vs. complex and they should simply strive to improve their DAG in terms of loading time, number of Tasks and interconnections between them.

  at memory usage, pay attention to the kind of memory you are observing. Usually you should look at
  ``working memory``(names might vary depending on your deployment) rather than ``total memory used``.

What can you do, to improve Scheduler's performance
"""""""""""""""""""""""""""""""""""""""""""""""""""

When you know what your resource usage is, the improvements that you can consider might be:

* improve the logic, efficiency of parsing and reduce complexity of your DAG Python code. It is parsed
  continuously so optimizing that code might bring tremendous improvements, especially if you try
  to reach out to some external databases etc. while parsing DAGs (this should be avoided at all cost).
  The :doc:`/best-practices` document shows a few examples on how you can approach dynamic DAG parsing
  without reaching out to external sources.
* improve utilization of your resources. This is when you have a free capacity in your system that
  seems underutilized (again CPU, memory I/O, networking are the prime candidates) - you can take
  actions like increasing number of schedulers, parsing processes or decreasing intervals for more

>   frequent actions might bring improvements in performance at the expense of higher utilization of those.

Reviewer 3: Do we offer some way of identifying these?
For example I use a test to check how much time it takes to load the dag and from that I deduce if an expensive call was made in operator init.

Author: Yeah. That's the https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#testing-a-dag which I will refer to.

  Microsoft SQLServer has not been tested with HA.

Fine-tuning your Scheduler performance
--------------------------------------

What impacts scheduler's performance
""""""""""""""""""""""""""""""""""""

The Scheduler is responsible for two operations:

* continuously parsing DAG files and synchronizing with the DAG in the database
* continuously scheduling tasks for execution

Those two tasks are executed in parallel by the scheduler and run independently of each other in
different processes. In order to fine-tune your scheduler, you need to include a number of factors:

* The kind of deployment you have

>     * what kind of filesystem you have to share the DAGs (impacts performance of continuous reading DAGs)

Reviewer: ```suggestion
    * what kind of filesystem you have to share the DAGs (impacts performance of continuously reading DAGs)
```

    * how fast the filesystem is (in many cases of distributed cloud filesystem you can pay extra to get
      more throughput/faster filesystem
    * how much memory you have for your processing
    * how much CPU you have available
    * how much networking throughput you have available

* The logic and definition of your DAG structure:
    * how many DAG files you have
    * how many DAGs you have in your files

>     * how large the DAG files are (remember scheduler needs to read and parse the file every n seconds)

Reviewer 4: ```suggestion
    * how large the DAG files are (remember dag parser needs to read and parse the file every n seconds)
```

    * how complex they are (i.e. how fast they can be parsed, how many tasks and dependencies they have)

>     * whether parsing your DAGs involves heavy processing (Hint! It should not. See :doc:`/best-practices`)

Reviewer 4: ```suggestion
    * whether parsing your DAG file involves heavy processing at the top level (Hint! It should not. See :ref:`best-practices/top_level_code`)
```

* The scheduler configuration
   * How many schedulers you have
   * How many parsing processes you have in your scheduler
   * How much time scheduler waits between re-parsing of the same DAG (it happens continuously)
   * How many task instances scheduler processes in one loop
   * How many new DAG runs should be created/scheduled per loop

>    * Whether to execute "mini-scheduler" after completed task to speed up scheduling dependent tasks

Reviewer 4: I think we shouldn't mention this -- as modulo bugs that we've fixed this is _awlays_ going to be better as it distributes the work, and also performs scheduling on a smaller subset of the task graph!

```suggestion
```

Author: Agree,

   * How often the scheduler should perform cleanup and check for orphaned tasks/adopting them

>    * Whether scheduler uses row-level locking

Reviewer 4: I'm not sure this should be mentioned, at least not without a massive warning -- turning it off means 1 scheduler only (or else incorrect behaviour!)

Author: Yeah. I think "no lock" is a bad idea (of mine) and we should remove it. It might have some undesireable effects

In order to perform fine-tuning, it's good to understand how Scheduler works under-the-hood.

> You can take a look at the ``Airflow Summit 2021``

Reviewer 4: ```suggestion
You can take a look at the Airflow Summit 2021 talk
```

Generally for fine-tuning, your approach should be the same as for any performance improvement and
optimizations (we will not recommend any specific tools - just use the tools that you usually use
to observe and monitor your systems):

* its extremely important to monitor your system with the right set of tools that you usually use to
  monitor your system. This document does not go into details of particular metrics and tools that you
  can use, it just describes what kind of resources you should monitor, but you should follow your best
  practices for monitoring to grab the right data.
* decide which aspect of performance is most important for you (what you want to improve)
* observe your system to see where your bottlenecks are: CPU, memory, I/O are the usual limiting factors
* based on your expectations and observations - decide what is your next improvement and go back to
  the observation of your performance, bottlenecks. Performance improvement is an iterative process.

What resources might limit Scheduler's performance
""""""""""""""""""""""""""""""""""""""""""""""""""

There are several areas of resource usage that you should pay attention to:

> * FileSystem performance. Airflow Scheduler relies heavily on parsing (sometimes a lot) of Python

Reviewer 4: This is less of a factor to scheduler overall, so I think we should move this further down the list.

Author: I think It does make a lot of difference (at least in perception of performance). There are really (mostly anecdotal but I saw it several time) evidences of how  - for example buing extra EFS IOPS improved both perceived performance and stability of scheduler. I've seen many people who believe that by employing a distributed filesystem, they "magically" got  instantly distributed files, which - especially with cloud Filesystem like EFS is not at all true. Also when working with Composer - they used GCS fuse under the hood and it caused a looot of troubles and stability/perceived performance issues before it was properly optimized and fine-tuned - and even then there were cases which make it work terribly (for example changing one line in a 100s of huge DAG files will require to effectively download all of those files from scratch and it might take minutes, causing all kind of problems where the files are in inconsistent state. 

I think we need to make people aware that choosing the right Filesystem matters and that it has huge impact on perceived  performance in a number of cases.

* CPU usage is most important for FileProcessors - those are the processes that parse and execute
  Python DAG files. Since Schedulers triggers such parsing continuously, when you have a lot of DAGs,
  the processing might take a lot of CPU. You can mitigate it by decreasing the
  :ref:`config:scheduler__min_file_process_interval`, but this is one of the mentioned trade-offs,
  result of this is that changes to such files will be picked up slower and you will see delays between
  submitting the files and getting them available in Airflow UI and executed by Scheduler. Optimizing
  the way how your DAGs are built, avoiding external data sources is your best approach to improve CPU
  usage. If you have more CPUs available, you can increase number of processing threads
  :ref:`config:scheduler__parsing_processes`, Also Airflow Scheduler scales almost linearly with
  several instances, so you can also add more Schedulers if your Scheduler's performance is CPU-bound.
* Airflow might use quite significant amount of memory when you try to get more performance out of it.
  Often more performance is achieved in Airflow by increasing number of processes handling the load,

>   and each process requires whole interpreter of Python loaded, a lot of classes imported, temporary

Reviewer 4: Because of forking and copy-on-write, each process doesn't pay the full cost of this memory usage: it is shared between processes,

i.e. one process might have 300Mb of active memory, but a second process doesn't need an extra 300Mb, it needs _almost_ nothing extra apart from the data it processes in memory.

Author: Agree in general for all the "shared C libraries" and "shared python libraries imported before fork" part. I will reword it to explain it better. It's indeed all about Heap Space and Stack under the hood, but it also depends a bit on a sequence of loading - for example with the "Optimized Hook imports" on one hand you have faster startup, but on the other hand, the lirbraries that are imported after the fork are duplicated - so the truth is somewhere in-between.

  in-memory storage. This can lead to memory pressure. You need to observe if your system is not using

>   more memory than it has - which results with using swap disk, which dramatically decreases performance.

Reviewer: Since you are referring to the scenario of using more memory than you have, I think it's clearer to word this way?
```suggestion
  in-memory storage. This can lead to memory pressure. You need to observe if your system is using
  more memory than it has - which results with using swap disk, which dramatically decreases performance.
```

  linearly (unless the shared database or filesystem is a bottleneck).
* experiment with different values for the "scheduler tunables". Often you might get better effects by
  simply exchanging one performance aspect for another. For example if you want to decrease the
  CPU usage, you might increase file processing interval (but the result will be that new DAGs will
  appear with bigger delay). Usually performance tuning is the art of balancing different aspects.
* sometimes you change scheduler behaviour slightly (for example change parsing sort order)
  in order to get better fine-tuned results for your particular deployment.

.. _scheduler:ha:tunables:

Scheduler Configuration options
"""""""""""""""""""""""""""""""

The following config settings can be used to control aspects of the Scheduler.
However you can also look at other non-performance-related scheduler configuration parameters available at
:doc:`../configurations-ref` in ``[scheduler]`` section.

- :ref:`config:scheduler__max_dagruns_to_create_per_loop`

  This changes the number of DAGs that are locked by each scheduler when
  creating DAG runs. One possible reason for setting this lower is if you

>   have huge DAGs and are running multiple schedules, you won't want one

Reviewer 4: ```suggestion
  have huge DAGs (in the order of 10k+ tasks per DAG) and are running multiple schedulers, you won't want one
```

  :ref:`config:scheduler__scheduler_health_check_threshold`) any running or
  queued tasks that were launched by the dead process will be "adopted" and
  monitored by this scheduler instead.

- :ref:`config:scheduler__dag_dir_list_interval`
  How often (in seconds) to scan the DAGs directory for new files.

- :ref:`config:scheduler__file_parsing_sort_mode`

>   The scheduler will list and sort the dag files to decide the parsing order.

Reviewer: ```suggestion
  The scheduler will list and sort the DAG files to decide the parsing order.
```

- :ref:`config:scheduler__max_tis_per_query`
  The batch size of queries in the scheduling main loop. If this is too high, SQL query
  performance may be impacted by one or more of the following:

  - reversion to full table scan - complexity of query predicate
  - excessive locking

  Additionally, you may hit the maximum allowable query length for your db.
  Set this to 0 for no limit (not advised)

- :ref:`config:scheduler__min_file_process_interval`
  Number of seconds after which a DAG file is parsed. The DAG file is parsed every
  min_file_process_interval number of seconds. Updates to DAGs are reflected after
  this interval. Keeping this number low will increase CPU usage.

- :ref:`config:scheduler__parsing_processes`

>   The scheduler can run multiple processes in parallel to parse dags. This defines

Reviewer: ```suggestion
  The scheduler can run multiple processes in parallel to parse DAGs. This defines
```

  how many processes will run.

- :ref:`config:scheduler__processor_poll_interval`
  The number of seconds to wait between consecutive DAG file processing

- :ref:`config:scheduler__schedule_after_task_execution`
  Should the Task supervisor process perform a “mini scheduler” to attempt to schedule more tasks of
  the same DAG. Leaving this on will mean tasks in the same DAG execute quicker,

>   but might starve out other dags in some circumstances

Reviewer: ```suggestion
  but might starve out other DAGs in some circumstances
```

>   but might starve out other dags in some circumstances

Reviewer 2: What circumstances?

Author: Don;t know. took it from the reference. I think this might be a bit different PR to review description of those config variables - as this also goes into "airflow.cfg" and the .yaml with configuration.

  SchedulerJobs.

  This setting controls how a dead scheduler will be noticed and the tasks it
  was "supervising" get picked up by another scheduler. The tasks will stay
  running, so there is no harm in not detecting this for a while.

  When a SchedulerJob is detected as "dead" (as determined by
  :ref:`config:scheduler__scheduler_health_check_threshold`) any running or
  queued tasks that were launched by the dead process will be "adopted" and
  monitored by this scheduler instead.

- :ref:`config:scheduler__dag_dir_list_interval`
  How often (in seconds) to scan the DAGs directory for new files.

- :ref:`config:scheduler__file_parsing_sort_mode`
  The scheduler will list and sort the DAG files to decide the parsing order.

- :ref:`config:scheduler__max_tis_per_query`
  The batch size of queries in the scheduling main loop. If this is too high, SQL query
  performance may be impacted by one or more of the following:

>   - reversion to full table scan - complexity of query predicate

Reviewer 4: I don't think this is true -- it looks like it just applies a `LIMIT` to the query.

(i.e. I can't see anywhere the gets TIs and then passes `n` of these to TI.filter_for_tis)

Author: I took it from the original description - so I will remove it there too.

  - excessive locking

  Additionally, you may hit the maximum allowable query length for your db.
  Set this to 0 for no limit (not advised).

- :ref:`config:scheduler__min_file_process_interval`

>   Number of seconds after which a DAG file is parsed. The DAG file is parsed every

Reviewer 4: ```suggestion
  Number of seconds after which a DAG file is reparsed. The DAG file is parsed every
```

  min_file_process_interval number of seconds. Updates to DAGs are reflected after
  this interval. Keeping this number low will increase CPU usage.

- :ref:`config:scheduler__parsing_processes`

>   The scheduler can run multiple processes in parallel to parse DAGs. This defines

Reviewer 4: ```suggestion
  The scheduler can run multiple processes in parallel to parse DAG files. This defines
```

  how many processes will run.

- :ref:`config:scheduler__processor_poll_interval`

>   The number of seconds to wait between consecutive DAG file processing.

Reviewer 4: This parameter is now badly named, as this is nothing to do with file processing anymore. Whoops.

This controls how long the scheduler will sleep between loops, but _iff_ there was nothing to do in the loop. i.e. if it scheduled something then it will start the next loop iteration straight away.

Author: I will explain it (and we might want to rename the parameter separately)
